datalad/interface/utils.py

```diff
@@ -525,24 +525,53 @@ def get_paths_by_dataset(paths, recursive=False, recursion_limit=None,
             d = dirname(path)
             if not d:
                 d = curdir
-        # this could be `None` if there is no git repo
-        dspath = dir_lookup.get(d, get_dataset_root(d))
-        dir_lookup[d] = dspath
+
+        dspath = dir_lookup.get(d, None)
+        if dspath:
+            _ds_looked_up = True
+        else:
+            _ds_looked_up = False
+            # this could be `None` if there is no git repo
+            dspath = get_dataset_root(d)
+            dir_lookup[d] = dspath
+
         if not dspath:
             nondataset_paths.append(path)
             continue
+
+        if path in out.get(dspath, []):
+            # we already recorded this path in the output
+            # this can happen, whenever `path` is a subdataset, that was
+            # discovered via recursive processing of another path before
+            continue
+
         if isdir(path):
             ds = Dataset(dspath)
             # we need to doublecheck that this is not a subdataset mount
-            # point, in which case get_toppath() would point to the parent
-            smpath = ds.get_containing_subdataset(
-                path, recursion_limit=1).path
-            if smpath != dspath:
-                # fix entry
-                dir_lookup[d] = smpath
-                # submodule still needs to be obtained
-                unavailable_paths.append(path)
-                continue
+            # point, in which case get_dataset_root() would point to the parent.
+
+            if not _ds_looked_up:
+                # we didn't deal with it before
+
+                smpath = ds.get_containing_subdataset(
+                    path, recursion_limit=1).path
+                if smpath != dspath:
+                    # fix entry
+                    dir_lookup[d] = smpath
+                    # submodule still needs to be obtained
+                    unavailable_paths.append(path)
+                    continue
+            else:
+                # we figured out the dataset previously, so we can spare some
+                # effort by not calling ds.subdatasets or
+                # ds.get_containing_subdataset. Instead we just need
+                # get_dataset_root, which is cheaper
+                if dspath != get_dataset_root(d):
+                    # if the looked up path isn't the default value,
+                    # it's a 'fixed' entry for an unavailable dataset (see above)
+                    unavailable_paths.append(path)
+                    continue
```

Comment:
note that above logic is not covered by tests and tests got one consistent failure already
NB didn't try to follow the logic yet... would prefer it first to be fixed/tested